specs/dynamicdependencies/DynamicDependencies.md

```diff
@@ -0,0 +1,1409 @@
+    /// a resolved package and its dependencies are added to the package graph.
+    Boolean OnlyUseFirstPackageFamily;
+
+    /// If a package is added to a package graph with a package of the same rank (aka a collision on rank)
+    /// and this option is true the package is prepended to the set of packages of the same rank.
+    /// By default packages are appended to the set of packages with the same rank.
+    Boolean PrependIfRankCollision;
+}
+
+/// TBD
+runtimeclass PackageDependency
+{
+    /// Create an intstance of the package dependency identified by id.
+    PackageDependency(String id);
+
+    /// Return the package dependency id.
+    String Id { get; }
+
+    /// Return true if the package dependencies would produce the same
+    /// package when resolved e.g. whether they share the same packageFamilyName,
+    /// minVersion, and packageDependencyProcessorArchitectures values.
+    static Boolean AreEquivalent(String packageDependencyId1, String packageDependencyId2);
+
+    /// Return true if this and other would produce the same
+    /// package when resolved e.g. whether they share the same packageFamilyName,
+    /// minVersion, and packageDependencyProcessorArchitectures values.
+    Boolean IsEquivalent(PackageDependency other);
+
+    /// Define a package dependency for the current user. The criteria for a PackageDependency
+    /// (package family name, minimum version, etc) may match multiple
+    /// packages, but ensures Deployment won't remove a package if it's
+    /// the only one satisfying the PackageDependency.
+    ///
+    /// @note A package matching a PackageDependency pin can still be removed
+    ///       as long as there's another that satisfies the PackageDependency.
+    ///       For example, if Fwk-v1 is installed and a PackageDependency specifies
+    ///       MinVersion=1 and then Fwk-v2 is installed, Deployment could remove
+    ///       Fwk-v1 because Fwk-v2 will satisfy the PackageDependency. After Fwk-v1
+    ///       is removed Deployment won't remove Fwk-v2 because it's the only package
+    ///       satisfying the PackageDependency. Thus  Fwk-v1 and Fwk-v2 (and any other
+    ///       package matching the PackageDependency) are 'loosely pinned' â€“ Deployment
+    ///       guarantees it won't remove a package if it would make a PackageDependency
+    ///       unsatisfied.
+    ///
+    /// A PackageDependency specifies criteria (package family, minimum version, etc)
+    /// and not a specific package. Deployment reserves the right to use a different
+    /// package (e.g. higher version) to satisfy the PackageDependency if/when
+    /// one becomes available.
+    ///
+    /// This method is equivalent to PinForUser(null,...).
+    ///
+    /// @param packageFamilyName the package family to pin
+    /// @param minVerrsion the minimum version to pin
+    ///
+    /// @note This fails if the package dependency cannot be resolved to a specific package.
```

Comment:
Fails how? By returning `null`? By throwing an exception? (We try not to throw exceptions for errors outside the app's control.)